elastic/apm-server
# txmetrics: Add `10m`, `1h` aggregation intervals

## Motivation/summary


Modifies the existing `txmetrics` aggregator to support a list of extra
aggregation intervals by introducing a new `RollUpIntervals` (internal
only) config option. The configuration option is not user facing, but
rather allow us to configure smaller intervals for testing purposes.
Also removes the `aggregation.transactions.interval` config option.

Two new intervals are added and published by default, 10 and 60m (1h).
and these are flushed when the APM Server is stopped or restarted.

Introduces a new `metrics-apm.transaction.${interval}m` datastream
which contains the mapping definitions for transaction metrics with a
series of changes to the apmpackage generator to allow interval
templating and individual ILM policies to be tied to each interval.
A new `Metricset.Interval` field is added to the ingested documents, to
allow them to contain the aggregated interval they represent. The field
format is `${interval}m` or `${interval}s` if < 60s.

Older APM Servers sending data to an Elasticsearch cluster with updated
APM Package, will continue indexing metrics to `metrics-apm.internal`.

## Checklist


## How to test these changes

1. `tilt up`
2. Run the `test.sh` script in https://gist.github.com/marclop/01a8f4270cc442573fd8319ca140015d.
3. Toggle expvar in the APM integration so metrics are published immediately on hot restart.
4. Verify that 3 metric sets per application and data stream `metrics-apm.transaction.1m`, `metrics-apm.transaction.10m`, `metrics-apm.transaction.60m`, and that they contain `event.duration` in nanoseconds matching the data stream interval (`constant_keyword`).
```
GET /metrics-apm*/_search
{
  "_source": [false],
  "fields": [
    "metricset.name",
    "event.duration",
    "service.name",
    "metricset.interval"
  ]
}
```
5. Verify that each of the different data streams has a dedicated ILM policy.

## Related issues

Part of #9703
Depends on https://github.com/elastic/apm-data/pull/7
